desc_source.go

```diff
@@ -302,3 +294,75 @@ func addFilesToSet(allFiles []*descriptorpb.FileDescriptorProto, expanded map[st
 	}
 	return append(allFiles, fd.AsFileDescriptorProto())
 }
+
+// WriteProtoFiles will use the given descriptor source to resolve all the given
+// symbols and write proto files with their definitions to the given output directory.
+func WriteProtoFiles(outProtoDirPath string, descSource DescriptorSource, symbols ...string) error {
+	filenames, fds, err := getFileDescriptors(symbols, descSource)
+	if err != nil {
+		return err
+	}
+	// now expand that to include transitive dependencies in topologically sorted
+	// order (such that file always appears after its dependencies)
+	expandedFiles := make(map[string]struct{}, len(fds))
+	allFilesSlice := make([]*desc.FileDescriptor, 0, len(fds))
+	for _, filename := range filenames {
+		allFilesSlice = addFilesToFileDescriptorList(allFilesSlice, expandedFiles, fds[filename])
+	}
+	pr := protoprint.Printer{}
+	// now we can serialize to files
+	for _, fd := range allFilesSlice {
+		fdFQName := fd.GetFullyQualifiedName()
+		dirPath := filepath.Dir(fdFQName)
+		outFilepath := filepath.Join(outProtoDirPath, dirPath)
+		if err := os.MkdirAll(outFilepath, 0755); err != nil {
+			return fmt.Errorf("failed to create directory %q: %v", outFilepath, err)
+		}
+		fileName := filepath.Base(fdFQName)
+		filePath := filepath.Join(outFilepath, fileName)
+		f, err := os.Create(filePath)
+		if err != nil {
+			if f != nil {
+				_ = f.Close()
+			}
+			return fmt.Errorf("failed to create file %q: %v", filePath, err)
+		}
+		if err := pr.PrintProtoFile(fd, f); err != nil {
+			_ = f.Close()
+			return fmt.Errorf("failed to write file %q: %v", filePath, err)
+		}
+		_ = f.Close()
```

Comment:
You actually want to error check the close as well.  Something like:
```go
err = pr.PrintProtoFile(fd, f)
if err == nil {
  err = f.Close()
}
// do the error check now
```
you also don't need the final close if you deferred close earlier

Reply:
ready